File: Napakalaki/src/napakalaki/Napakalaki.java
/**
 * 
 */

package napakalaki;

import java.util.ArrayList;

public class Napakalaki {
    //Datos Miembros
    private static final Napakalaki instance = new Napakalaki();

    private Player currentPlayer;
    private int currentIndexPlayer;
    private Monster currentMonster;
    private ArrayList<Player> players;

    /*Constructor (es privado para que la clase sea singleton), no conocemos la
    implementación del constructor probablemente inicie a los players a traves 
    de una peticion, y posteriormente inicie el juego con initgame() por ello:
    */
    private Napakalaki(){
         Player currentPlayer = null;
         currentIndexPlayer = -1;
         currentMonster = null;
         players = null;
    }
    /*
    Inicializa la lista de jugadores a partir de sus nombres
    Ademas del indice de primer jugador y el jugador actual.
    */
    private void initPlayers(ArrayList<String> names){
        players = new ArrayList<>();
        if(names.size()>0){
            for(int i=0; i<names.size(); i++){
                Player newPlayer = new Player(names.get(i));
                newPlayer.initTreasures();
                players.add(newPlayer);
            }
            currentIndexPlayer = Dice.getInstance().nextNumber()%names.size();
            currentPlayer = players.get(currentIndexPlayer);
            currentMonster= CardDealer.getInstance().nextMonster();
        }
        else{
            System.out.println("DEPURACION DE ERROR: Lista de nombres vacia");
        }
    }
    
    public void initGame(ArrayList<String> players){
        CardDealer dealer = CardDealer.getInstance();
        dealer.initCards(); //Inicializa y baraja
        initPlayers(players);
    }
        
    /* Actualiza el jugador actual y su indice, teniendo en cuenta que despues
    del ultimo jugador le toca de nuevo al primero.
    */
    private Player nextPlayer(){
        currentIndexPlayer = (currentIndexPlayer+1) % players.size();
        currentPlayer = players.get(currentIndexPlayer);
        return currentPlayer;
    }

    /*Combat devuelve el resultado del combate entre el jugador actual y
    el monstruo actual.
    */
    public CombatResult combat(){
        CombatResult stat; 
        stat = currentPlayer.combat(currentMonster);
        
        //En caso de perder y convertirse el jugador actual se convierte en cultist
        if(stat==CombatResult.LOSEANDCONVERT){
            Cultist cultistCard = CardDealer.getInstance().nextCultist();
            CultistPlayer cPlayer = new CultistPlayer(currentPlayer, cultistCard);
            currentPlayer = cPlayer;
            players.set(currentIndexPlayer, cPlayer);
        }
        return stat;
    }
    /*
    El jugador descarta el tesoro t de su lista de tesoros visibles
    */
    public void discardVisibleTreasure(Treasure t){
        currentPlayer.discardVisibleTreasure(t);
    }
    /*
    El jugador descarta el tesoro t de su lista de tesoros ocultos
    */
    public void discardHiddenTreasure(Treasure t){
        currentPlayer.discardHiddenTreasure(t);
    }
     /*
    El jugador se equipa con uno de los tesoros ocultos (t) y lo pasa a la lista 
    de tesoros visibles.
    Devuelve true en el caso de que se haya podido hacer visible.
    */
    public boolean makeTreasureVisible(Treasure t){
        boolean canM;
        if(canMakeTreasureVisible(t)){
            currentPlayer.makeTreasureVisible(t);
            canM=true;
        }
        else
            canM = false;
        return canM;
    }

    public boolean buyLevels(ArrayList<Treasure> visible, ArrayList<Treasure> hidden){
        boolean canI = currentPlayer.buyLevels(visible, hidden);
        return canI;
    }


    public Player getCurrentPlayer(){
        return currentPlayer;
    }

    public Monster getCurrentMonster(){
        return currentMonster;
    }

    public boolean canMakeTreasureVisible(Treasure t){
        boolean canM = currentPlayer.canMakeTreasureVisible(t);
        return canM;
    }

    public ArrayList<Treasure> getVisibleTreasures(){
        return currentPlayer.getVisibleTreasures();                         
    }

    public ArrayList<Treasure> getHiddenTreasures(){
        return currentPlayer.getHiddenTreasures();                          
    }

    public boolean nextTurn(){
        CardDealer cd = CardDealer.getInstance();
        boolean stateOK;                           
        boolean dead;
        stateOK = nextTurnAllowed();
        if(stateOK){
            currentMonster = cd.nextMonster();
            currentPlayer = nextPlayer();
            dead = currentPlayer.isDead();
            if(dead){
                currentPlayer.initTreasures();
            }
            return true;
        }
        else
            return false;
        
    }

    public boolean nextTurnAllowed(){
        return currentPlayer.validState();
    }
    public boolean endOfGame(CombatResult result){
        if(result == CombatResult.WINANDWINGAME){
            return true;
        }
        else
            return false;
    }

    //getter
    public static Napakalaki getInstance() {
        return instance;
    }
}

File: Napakalaki_Ruby/lib/napakalaki.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

  require_relative 'bad_consequence.rb'
  require_relative 'prize.rb'
  require_relative 'monster.rb'
  require_relative 'treasure_kind.rb'
  require_relative 'player.rb'
  require_relative 'combat_result.rb'
  require_relative 'card_dealer.rb'
  require "singleton"
  
module Napakalaki
  
    class Napakalaki
      
      include Singleton
      
      attr_reader :current_player, :current_monster, :players, :dealer
      
      private
      def init_players(names)
        @players = Array.new
        @dealer = CardDealer.instance()
        names.each() do |p|
          @players << Player.new(p) 
        end
      end
      
      def next_player()
        total_players = @players.size()-1
        if (@current_player == nil)
          next_index = rand(total_players)
        else
          current_player_index = @players.index(@current_player)
          if current_player_index == total_players
            next_index = 0
          else
            next_index = current_player_index + 1
          end
        end
        next_player = @players[next_index]
        player = next_player
        return player
      end
      
      def next_turn_allowed()        
        if @current_player == nil then
          allowed = true
        else
          allowed = @current_player.valid_state()
        end
        return allowed
      end
      
      public
      
      def develop_combat()
        combat = @current_player.combat(@current_monster)
        @dealer.give_monster_back(@current_monster)  
        if combat == CombatResult::LoseAndConvert
            @dealer = CardDealer.instance()
            cultist_card = dealer.next_cultist();
            c_player = new CultistPlayer.new(@current_player, cultist_card);
            @current_player = c_player;
            @players[@players.index(@current_player)] = c_player
        end
        return combat
      end
      
      def discard_visible_treasures(treasures)
        treasures.each() do |t|
          @current_player.discard_visible_treasure(t)
          @dealer.give_treasures_back(t)
        end
      end
      
      def discard_hidden_treasures(treasures)
        treasures.each() do |t|
          @current_player.discard_hidden_treasure(t)
          @dealer.give_treasures_back(t)
        end
      end
      
      def make_treasures_visible(treasures)
        treasures.each() do |t|
          @current_player.make_treasure_visible(t)
        end
      end
      
      def buy_levels(visible, hidden)
        return @current_player.buy_levels(visible, hidden)
      end
      
      def init_game(players)
        init_players(players)
        @dealer.init_cards()
        next_turn()
      end
            
      def next_turn()
        state = next_turn_allowed()
        if state
          @current_monster = @dealer.next_monster()
          @current_player = next_player()    
          if @current_player.is_dead()
            @current_player.init_treasures
          end
          else @current_monster = @dealer.next_monster()
        end
        return state
      end
      
      def end_of_game(result)
        if result == combat_result::WinAndWinGame
          return true
        else
          return false
        end
      end     
 
  end
end

